src/server/main_service.cc

```diff
@@ -882,7 +883,11 @@ void Service::DispatchCommand(CmdArgList args, facade::ConnectionContext* cntx)
     dfly_cntx->reply_builder()->CloseConnection();
   }
 
-  end_usec = ProactorBase::GetMonotonicTimeNs();
+  end_usec = ProactorBase::GetMonotonicTimeNs();  // misleading name, this is actually ns
+  auto after = absl::GetCurrentTimeNanos();
+
+  etl.cmd_sum_map[cid->name()] += (after - before) / 1000.0;
```

Comment:
you perform one lookup at line 818 and one here. 
1. Between both events the statistics are skewed because `cmd_calls` increased before.
2. You perform 2 hashtable lookups but we could do it in a single lookup by maintaining a single map with 
`Value = {.calls, .sum} `

Reply:
Kind of stumped on this!

The lookup on line 818 happens before a command is invoked/scheduled (scheduled being in MULTI) so incrementing the calls counter there is wrong. I can't see how to avoid two lookups if I correct this behavior and wish to track MULTI commands.

So the options are either having two lookups or not counting commands executed under MULTI to the total counter.

Reply:
I'm pretty fine with just skipping MULTI/EXEC for this metric, I don't know if there's a very useful way to define duration for them

Reply:
Yes, they are not counted towards the metric, however they can be counted towards total calls to some command. This would require another counter (say `total_calls`) which would need another lookup.

So this is a matter of `total_calls` vs two lookups.

Reply:
I do not follow. Where the second lookup should be, in your opinion?

Reply:
Line 818, before queuing MULTI commands

Reply:
have you checked what happens when you remove this line and run 
```
multi
PING
exec
```

Reply:
Commands executed under MULTI are counted towards the total number of calls, their running time is not.

Reply:
I still do not understand :)
line 889 increases the total number of calls.
line 890 increases the running time. 
how these metrics can diverge?

Reply:
can you post here the screenshot of `Commandstats` for this sequence?

Reply:
Ok, I'll rephrase the story: originally the increment happened [here](https://github.com/dragonflydb/dragonfly/blob/590bf94e1cdd7b30141f798e4a1cd3d219116914/src/server/main_service.cc#L818), before doing anything. This captured all calls to commands, including those queued.

My last commit, [d7b6346](https://github.com/dragonflydb/dragonfly/pull/1552/commits/d7b6346b2213fd3c08a104a8279fa91611bcd00e), moved the increment to *after running the command*, which captures all calls **except** calls inside MULTI. If we wish to count all calls, two lookups are necessary, otherwise the computation of the average wouldn't be accurate

Reply:
As for executing the sequence, here is the output from the main branch:
```
# Commandstats
cmd_COMMAND:1 # (redis-cli)
cmd_EXEC:1
cmd_INFO:1
cmd_MULTI:1
cmd_PING:1
```

And running from this branch:
```
# Commandstats
cmdstat_COMMAND:calls=1,usec=19428,usec_per_call=19428 # (redis-cli)
cmdstat_EXEC:calls=1,usec=354,usec_per_call=354
cmdstat_MULTI:calls=1,usec=193,usec_per_call=193
```

note that PING is missing.

Reply:
do you know how to use gdb?
using gdb it's easy to find that ping is being executed through multi_command_squasher code.
Essentially all command invocations happen via `cmd->Cid()->Invoke(args, cntx_);` calls. 
Therefore, I suggest moving latency and counts measurements inside this function.

`Invoke(..)` is defined `command_registry.h` . Please move it to .cc and measure the timings there. you can also update the stats there as well.